### PR TITLE
Add shard key encoding primitives for split planning

### PR DESCRIPTION
**Pull request by ahrav**
## Summary

Adds the foundational key-encoding module (`key_encoding.rs`) to `gossip-contracts::shard`, providing byte-level primitives that the split planner will use to compute shard boundaries. This is pure, self-contained key arithmetic with no side effects — it operates on `&[u8]` and `Vec<u8>` only.

**Impact:** 2 source files changed (1 new, 1 modified) — 568 net additions

## What Changed

**New file: `crates/gossip-contracts/src/shard/key_encoding.rs` (567 lines)**

| Component | Purpose |
|-----------|---------|
| `KeyEncoding` trait | Contract for connectors to encode typed keys into lex-ordered bytes |
| `prefix_successor(prefix)` | Exclusive upper bound for prefix scans (FoundationDB `strinc` equivalent) |
| `key_successor(key)` | Minimal strict successor respecting `MAX_KEY_SIZE` (4 KiB) |
| `byte_midpoint(a, b)` | Approximate bisection point for split planning |
| `PrefixShardError` enum | Error type for prefix-to-range conversion failures |

**Modified file: `crates/gossip-contracts/src/shard/mod.rs`**

- Replaced placeholder module docs with accurate description of current contents
- Added `pub mod key_encoding` and re-exports for all public items

## Design Decisions

1. **Functions return `Option`, not `Result`** — These are boundary arithmetic functions where `None` means "no valid result exists within representable bounds." The caller (split planner) treats `None` as "range too narrow to bisect."

2. **`key_successor` prefers append-zero over increment** — Appending `0x00` is infallible for keys below `MAX_KEY_SIZE` and produces the tightest possible successor. Increment (via `prefix_successor`) is the fallback only at the size ceiling.

3. **`byte_midpoint` uses canonicalization with bounds check** — Leading-zero trimming can push the midpoint outside `(a, b)` under lex ordering. The function returns `None` for these edge cases rather than padding to fixed width.

4. **`PrefixShardError` wraps `ShardSpecInputError`** — The error chain delegates to the existing `shard_spec` error type, keeping the hierarchy flat.

## Testing

27 tests passing, mixing three approaches:

| Type | Count | Coverage |
|------|-------|----------|
| Unit tests | 7 | Basic behavior, regression cases, error formatting/delegation |
| `rstest` parameterized | 2 | Edge cases for `byte_midpoint` (8 cases) and `PrefixShardError::Display` (3 cases) |
| `proptest` property tests | 8 | Ordering invariants, upper-bound proofs, successor/midpoint contracts over random inputs |

Key properties verified:
- `prefix_successor` result is always strictly greater than input
- `prefix_successor` is a valid upper bound for all keys sharing the prefix
- `byte_midpoint` result is always strictly between `a` and `b`
- `key_successor` returns `None` only for oversized keys or all-`0xFF` at max size
- `key_successor` delegates to `prefix_successor` at exactly `MAX_KEY_SIZE`

## Risk Assessment

**Low risk.** New, isolated module with no callers yet. Adds code but changes no existing behavior.

| Factor | Assessment |
|--------|-----------|
| Breaking changes | None — new public API only |
| Performance | Pure arithmetic on small byte slices |
| Security | No I/O, no parsing of untrusted input |
| Dependencies | Only `crate::coordination::shard_spec::{MAX_KEY_SIZE, ShardSpecInputError}` |

## Review Checklist

- [ ] `prefix_successor` algorithm: trailing `0xFF` stripping + increment
- [ ] `byte_midpoint` big-endian addition/halving: carry propagation and remainder handling
- [ ] `key_successor` strategy table: append-zero vs increment vs reject
- [ ] `PrefixShardError` variants cover all failure modes
- [ ] Property tests cover the ordering contracts claimed in documentation
- [ ] Module docs accurately describe scope boundaries

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/gossip-contracts/src/shard/key_encoding.rs [hunk 1]
```diff
@@ -0,0 +1,567 @@
...
+#[must_use]
+pub fn byte_midpoint(a: &[u8], b: &[u8]) -> Option<Vec<u8>> {
+    if a >= b {
+        return None;
+    }
+
+    let max_len = a.len().max(b.len());
+    if max_len == 0 {
+        return None;
+    }
```
[the remark this answers is not in the document]
Comment (ahrav) [reply crates/gossip-contracts/src/shard/key_encoding.rs]: Good catch. Confirmed — `byte_midpoint` accepted oversized inputs and produced an oversized result. Added a `max_len > MAX_KEY_SIZE` guard consistent with `key_successor`, plus two tests: one verifying oversized inputs are rejected, one verifying valid-sized inputs produce valid-sized output. Doc comment updated to document the new `None` condition.
